### PR TITLE
Make the header length match optional.

### PR DESCRIPTION
**Pull request by Tratcher**
#37 @davidfowl @muratg @pakrym

### CODE REVIEW

**Comment by pakrym** [comment]
:shipit:

File: src/Microsoft.AspNetCore.HttpOverrides/Internal/IPEndPointParser.cs [hunk 1]
```diff
@@ -14,6 +14,11 @@ public static bool TryParse(string addressWithPort, out IPEndPoint endpoint)
             IPAddress address;
             endpoint = null;
 
+            if (string.IsNullOrEmpty(addressWithPort))
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.HttpOverrides/Internal/IPEndPointParser.cs]: Is this new case covered in any test?

File: src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersMiddleware.cs [hunk 2]
```diff
@@ -65,75 +65,61 @@ public void ApplyForwarders(HttpContext context)
...
-                offset = entryCount - _options.ForwardLimit.Value;
                 entryCount = _options.ForwardLimit.Value;
             }
 
             // Group the data together.
             var sets = new List<SetOfForwarders>(entryCount);
             for (int i = 0; i < entryCount; i++)
             {
+                // They get processed in reverse order, right to left.
                 var set = new SetOfForwarders();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersMiddleware.cs]: Lets change SetOfForwarders to a struct

File: src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersMiddleware.cs [hunk 3]
```diff
@@ -65,75 +65,61 @@ public void ApplyForwarders(HttpContext context)
...
             // Apply ForwardLimit, if any
-            int offset = 0;
             if (_options.ForwardLimit.HasValue && entryCount > _options.ForwardLimit)
             {
-                offset = entryCount - _options.ForwardLimit.Value;
                 entryCount = _options.ForwardLimit.Value;
             }
 
             // Group the data together.
             var sets = new List<SetOfForwarders>(entryCount);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersMiddleware.cs]: Does it even need to be a list? Can't this just be an array?

File: src/Microsoft.AspNetCore.HttpOverrides/Internal/IPEndPointParser.cs (hunk 1 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpOverrides/Internal/IPEndPointParser.cs]: Yeah, I hit this when having more schemes than IPs.

**Comment by Tratcher** [comment]
@davidfowl Updated

**Comment by Tratcher** [comment]
Rebased and merged.

**Pull request closed without merging**